### PR TITLE
Add metrics interfaces at edge

### PR DESCRIPTION
**Pull request by fisherxu**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
Previously, edgecore did not support kubelet's metrics interface, and users could not view monitoring data at the edge.

This PR integrates cadvisor-based metrics interface into edged, and users can collect monitoring data.

**Notes**:  the metrics interface is now only accessed on localhost based insecure server now.

Next step, we will support metric-server in the cloud, and support collect monitoring data from the edge through the cloud/edge hub.

And the edgecore binary size change(Uncompressed): 
92M -> 105M

The memory change:
 (**0 pod**): 64M -> 72M  
 (**100 pods**): 72M ->96M 

Disable metrics through setting `EnableMetric` to false:
 (**0 pod**): 64M   
(**100 pods**): 80M

Xref: #1561

```
➜  ~ curl 127.0.0.1:10350/metrics/probes   
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.58564475334e+09
```

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Add metrics interfaces at edge
```

### CODE REVIEW

**Comment by fisherxu** [comment]
Metrics-server can use `stats/summary` below to collect infos.
```
➜  ~ kubectl get pods
NAME                                READY   STATUS    RESTARTS   AGE
nginx-deployment-86b86956b8-vb868   1/1     Running   0          15s
➜  ~ 
➜  ~ curl "127.0.0.1:10350/stats/summary"
{
 "node": {
  "nodeName": "edge-node",
  "systemContainers": [
   {
    "name": "kubelet",
    "startTime": "2020-04-03T10:16:05Z",
    "cpu": {
     "time": "2020-04-03T10:19:14Z",
     "usageNanoCores": 97041104,
     "usageCoreNanoSeconds": 1421531396180
    },
    "memory": {
     "time": "2020-04-03T10:19:14Z",
     "usageBytes": 525459456,
     "workingSetBytes": 468848640,
     "rssBytes": 74661888,
     "pageFaults": 45787751,
     "majorPageFaults": 1418
    }
   },
   {
    "name": "runtime",
    "startTime": "2020-04-03T10:16:05Z",
    "cpu": {
     "time": "2020-04-03T10:19:15Z",
     "usageNanoCores": 159960549,
     "usageCoreNanoSeconds": 163277825011
    },
    "memory": {
     "time": "2020-04-03T10:19:15Z",
     "usageBytes": 1162891264,
     "workingSetBytes": 1005391872,
     "rssBytes": 362520576,
     "pageFaults": 3391977,
     "majorPageFaults": 0
    }
   },
   {
    "name": "pods",
    "startTime": "2020-04-02T08:13:38Z",
    "cpu": {
     "time": "2020-04-03T10:19:06Z",
     "usageNanoCores": 356500246,
     "usageCoreNanoSeconds": 853667101402041
    },
    "memory": {
     "time": "2020-04-03T10:19:06Z",
     "availableBytes": 2506276864,
     "usageBytes": 7189700608,
     "workingSetBytes": 5864239104,
     "rssBytes": 1310015488,
     "pageFaults": 733896478,
     "majorPageFaults": 270
    }
   }
  ],
  "startTime": "2019-12-23T11:54:27Z",
  "cpu": {
   "time": "2020-04-03T10:19:06Z",
   "usageNanoCores": 356500246,
   "usageCoreNanoSeconds": 853667101402041
  },
  "memory": {
   "time": "2020-04-03T10:19:06Z",
   "availableBytes": 2506276864,
   "usageBytes": 7189700608,
   "workingSetBytes": 5864239104,
   "rssBytes": 1310015488,
   "pageFaults": 733896478,
   "majorPageFaults": 270
  },
  "network": {
   "time": "2020-04-03T10:19:06Z",
   "name": "eth0",
   "rxBytes": 22677920837,
   "rxErrors": 0,
   "txBytes": 11514467359,
   "txErrors": 0,
   "interfaces": [
    {
     "name": "edge0",
     "rxBytes": 0,
     "rxErrors": 0,
     "txBytes": 0,
     "txErrors": 0
    },
    {
     "name": "dummy0",
     "rxBytes": 0,
     "rxErrors": 0,
     "txBytes": 0,
     "txErrors": 0
    },
    {
     "name": "cni0",
     "rxBytes": 6397556,
     "rxErrors": 0,
     "txBytes": 14659853,
     "txErrors": 0
    },
    {
     "name": "flannel.1",
     "rxBytes": 0,
     "rxErrors": 0,
     "txBytes": 0,
     "txErrors": 0
    },
    {
     "name": "eth0",
     "rxBytes": 22677920837,
     "rxErrors": 0,
     "txBytes": 11514467359,
     "txErrors": 0
    }
   ]
  },
  "fs": {
   "time": "2020-04-03T10:19:06Z",
   "availableBytes": 47825649664,
   "capacityBytes": 126692048896,
   "usedBytes": 73266397184,
   "inodesFree": 6326802,
   "inodes": 7864320,
   "inodesUsed": 1537518
  },
  "runtime": {
   "imageFs": {
    "time": "2020-04-03T10:19:06Z",
    "availableBytes": 47825649664,
    "capacityBytes": 126692048896,
    "usedBytes": 7693978269,
    "inodesFree": 6326802,
    "inodes": 7864320,
    "inodesUsed": 1537518
   }
  },
  "rlimit": {
   "time": "2020-04-03T10:19:22Z",
   "maxpid": 32768,
   "curproc": 3124
  }
 },
 "pods": [
  {
   "podRef": {
    "name": "nginx-deployment-86b86956b8-vb868",
    "namespace": "default",
    "uid": "5911a297-175e-431b-8c22-3b85586aaa12"
   },
   "startTime": "2020-04-03T10:19:04Z",
   "containers": [
    {
     "name": "nginx",
     "startTime": "2020-04-03T10:19:08Z",
     "cpu": {
      "time": "2020-04-03T10:19:19Z",
      "usageNanoCores": 0,
      "usageCoreNanoSeconds": 29694404
     },
     "memory": {
      "time": "2020-04-03T10:19:19Z",
      "availableBytes": 131338240,
      "usageBytes": 2887680,
      "workingSetBytes": 2879488,
      "rssBytes": 1441792,
      "pageFaults": 1114,
      "majorPageFaults": 0
     },
     "rootfs": {
      "time": "2020-04-03T10:19:19Z",
      "availableBytes": 47825649664,
      "capacityBytes": 126692048896,
      "usedBytes": 57344,
      "inodesFree": 6326802,
      "inodes": 7864320,
      "inodesUsed": 14
     },
     "logs": {
      "time": "2020-04-03T10:19:19Z",
      "availableBytes": 47825649664,
      "capacityBytes": 126692048896,
      "usedBytes": 24576,
      "inodesFree": 6326802,
      "inodes": 7864320,
      "inodesUsed": 1537518
     }
    }
   ],
   "network": {
    "time": "2020-04-03T10:19:15Z",
    "name": "eth0",
    "rxBytes": 648,
    "rxErrors": 0,
    "txBytes": 0,
    "txErrors": 0,
    "interfaces": [
     {
      "name": "eth0",
      "rxBytes": 648,
      "rxErrors": 0,
      "txBytes": 0,
      "txErrors": 0
     }
    ]
   },
   "volume": [
    {
     "time": "2020-04-03T10:19:05Z",
     "availableBytes": 4185243648,
     "capacityBytes": 4185255936,
     "usedBytes": 12288,
     "inodesFree": 1021782,
     "inodes": 1021791,
     "inodesUsed": 9,
     "name": "default-token-qnbp2"
    }
   ],
   "ephemeral-storage": {
    "time": "2020-04-03T10:19:19Z",
    "availableBytes": 47825649664,
    "capacityBytes": 126692048896,
    "usedBytes": 81920,
    "inodesFree": 6326802,
    "inodes": 7864320,
    "inodesUsed": 14
   }
  }
 ]
}#                              
```

**Comment by fisherxu** [comment]
It's ready to review now, ptal 
@kevin-wangzefeng @kadisi @subpathdev @chendave

File: edge/pkg/edged/cadvisor/cadvisor_linux.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
+// +build cgo,linux
```
Comment (kevin-wangzefeng) [inline edge/pkg/edged/cadvisor/cadvisor_linux.go]: This is not actually depending on linux, maybe remove _linux in filename, and remove corresponding build arg

File: edge/pkg/edged/edged_pods.go [hunk 2]
```diff
@@ -1150,3 +1151,165 @@ func (e *edged) getHostIPByInterface() (string, error) {
...
+	}
+	// Resolve and type convert back again.
+	// We need the static pod UID but the kubecontainer API works with types.UID.
+	podUID = types.UID(e.podManager.TranslatePodUID(podUID))
+	pod := kubecontainer.Pods(pods).FindPod(podFullName, podUID)
+	return pod.FindContainerByName(containerName), nil
+}
+
+// RunInContainer runs a command in a container, returns the combined stdout, stderr as an array of bytes
+func (e *edged) RunInContainer(podFullName string, podUID types.UID, containerName string, cmd []string) ([]byte, error) {
```
Comment (chendave) [inline edge/pkg/edged/edged_pods.go:1170]: where does this method got called?

File: edge/pkg/edged/edged_pods.go [hunk 3]
```diff
@@ -1150,3 +1151,165 @@ func (e *edged) getHostIPByInterface() (string, error) {
...
+		return nil, fmt.Errorf("container not found (%q)", containerName)
+	}
+	// TODO(tallclair): Pass a proper timeout value.
+	return e.runner.RunInContainer(container.ID, cmd, 0)
+}
+
+// GetKubeletContainerLogs returns logs from the container
+// TODO: this method is returning logs of random container attempts, when it should be returning the most recent attempt
+// or all of them.
+func (e *edged) GetKubeletContainerLogs(ctx context.Context, podFullName, containerName string, logOptions *v1.PodLogOptions, stdout, stderr io.Writer) error {
```
Comment (chendave) [inline edge/pkg/edged/edged_pods.go:1185]: Does this has anything to do with the metrics?

File: edge/pkg/edged/edged_pods.go [hunk 4]
```diff
@@ -1150,3 +1157,165 @@ func (e *edged) getHostIPByInterface() (string, error) {
...
+	return nil, nil
+}
+
+func (e *edged) GetAttach(podFullName string, podUID types.UID, containerName string, streamOpts remotecommand.Options) (*url.URL, error) {
+	return nil, nil
+}
+
+func (e *edged) GetPortForward(podName, podNamespace string, podUID types.UID, portForwardOpts portforward.V4Options) (*url.URL, error) {
+	return nil, nil
+}
```
Comment (chendave) [inline edge/pkg/edged/edged_pods.go:1249]: basically, I hope this PR focus on metrics only.

File: edge/pkg/edged/server/server.go [hunk 5]
```diff
@@ -14,7 +14,7 @@ import (
 //constants to define server address
 const (
 	ServerAddr = "127.0.0.1"
-	ServerPort = "10255"
+	ServerPort = "10350"
```
Comment (chendave) [inline edge/pkg/edged/server/server.go:17]: why?

File: edge/pkg/edged/edged.go [hunk 6]
```diff
@@ -581,8 +581,8 @@ func newEdged(enable bool) (*edged, error) {
 		ed.cadvisor,
 		cm.NodeConfig{
 			CgroupDriver:                 edgedconfig.Config.CGroupDriver,
-			SystemCgroupsName:            edgedconfig.Config.CGroupDriver,
-			KubeletCgroupsName:           edgedconfig.Config.CGroupDriver,
+			SystemCgroupsName:            "",
+			KubeletCgroupsName:           "",
```
Comment (chendave) [inline edge/pkg/edged/edged.go:585]: better to be defined as config options?

File: edge/pkg/edged/edged_pods.go [hunk 7]
```diff
@@ -554,13 +552,9 @@ func (e *edged) removeOrphanedPodStatuses(pods []*v1.Pod) {
...
-	ret := e.cgroupDriver
-	if edgedconfig.Config.RemoteRuntimeEndpoint == DockerShimEndpoint ||
-		edgedconfig.Config.RemoteRuntimeEndpoint == DockerShimEndpointDeprecated {
-		//always have a ".slice" suffix
-		ret = ret + systemdSuffix
-	}
-	return ret
+	pcm := e.containerManager.NewPodContainerManager()
+	_, cgroupParent := pcm.GetPodContainerName(pod)
+	return cgroupParent
```
Comment (chendave) [inline edge/pkg/edged/edged_pods.go:557]: those piece has been merged.

File: edge/pkg/edged/edged_pods.go (hunk 4 above)
Comment (kadisi) [reply edge/pkg/edged/edged_pods.go:1249]: These methods are to implement the `HostInterface` interface

i think this pr is not very easy to split.

File: edge/pkg/edged/server/server.go (hunk 5 above)
Comment (kadisi) [reply edge/pkg/edged/server/server.go:17]: This is to support both the logs and exec feature 。 #1477

File: edge/pkg/edged/edged_pods.go (hunk 2 above)
Comment (fisherxu) [reply edge/pkg/edged/edged_pods.go:1170]: Need implement the HostInterface interface in edged for the handler.

File: edge/pkg/edged/edged_pods.go (hunk 3 above)
Comment (fisherxu) [reply edge/pkg/edged/edged_pods.go:1185]: ditto

File: edge/pkg/edged/edged_pods.go (hunk 4 above)
Comment (fisherxu) [reply edge/pkg/edged/edged_pods.go:1249]: ditto

File: edge/pkg/edged/edged_pods.go (hunk 7 above)
Comment (fisherxu) [reply edge/pkg/edged/edged_pods.go:557]: Will rebased.

File: edge/pkg/edged/cadvisor/cadvisor_linux.go (hunk 1 above)
Comment (fisherxu) [reply edge/pkg/edged/cadvisor/cadvisor_linux.go]: Done

File: edge/pkg/edged/edged.go [hunk 8]
```diff
@@ -343,6 +359,28 @@ func getRuntimeAndImageServices(remoteRuntimeEndpoint string, remoteImageEndpoin
 	return rs, is, err
 }
 
+func (e *edged) cgroupRoots() []string {
+	var cgroupRoots []string
+
+	cgroupRoots = append(cgroupRoots, cm.NodeAllocatableRoot("", edgedconfig.Config.CGroupDriver))
```
[the remark this answers is not in the document]
Comment (fisherxu) [reply edge/pkg/edged/edged.go:366]: Done

File: edge/pkg/edged/server/server.go (hunk 5 above)
Comment (fisherxu) [reply edge/pkg/edged/server/server.go:17]: Yes, we can defined as config options later along with cloud side.

File: edge/pkg/edged/edged.go (hunk 6 above)
Comment (fisherxu) [reply edge/pkg/edged/edged.go:585]: These two fields now have no clear purpose on the edge side, let's expose it if needed later.

Comment (fisherxu) [reply edge/pkg/edged/edged.go]: Finally, exposed them ;-)

File: edge/pkg/edged/edged.go [hunk 9]
```diff
@@ -526,10 +584,42 @@ func newEdged(enable bool) (*edged, error) {
...
 	ed.containerManager = containerManager
 	ed.runtimeService = runtimeService
+
+	runtimeCache, err := kubecontainer.NewRuntimeCache(ed.containerRuntime)
+	if err != nil {
+		return nil, err
+	}
+	ed.runtimeCache = runtimeCache
+
+	ed.resourceAnalyzer = serverstats.NewResourceAnalyzer(ed, time.Minute)
```
[the remark this answers is not in the document]
Comment (fisherxu) [reply edge/pkg/edged/edged.go]: Done.

**Comment by fisherxu** [comment]
Pr has rebased, ptal @kevin-wangzefeng @chendave @kadisi

**Comment by kadisi** [comment]
/lgtm

thanks  @fisherxu

**Pull request merged**